--- a/configure	2017-04-06 14:48:19.827916572 +0200
+++ nrpe-3.0.1/configure	2016-09-08 18:18:58.000000000 +0200
@@ -3037,7 +3037,7 @@
   aix|hp-ux) :
     inetd_type=`UNIX95= ps -A -o comm | grep inetd | head -1` ;; #(
   *) :
-    inetd_type=`ps -C "inetd,xinetd" -o fname | grep -vi COMMAND` ;; #(
+    inetd_type=`ps -C "inetd,xinetd" -o fname | grep -vi COMMAND | head -1` ;; #(
   *) :
      ;;
 esac
