From ef4e15d881a1f7d4e93309cbf293eeb3021410a6 Mon Sep 17 00:00:00 2001
From: Danijel Tasov <data@consol.de>
Date: Fri, 20 Oct 2017 15:47:09 +0200
Subject: [PATCH] warn if no volumes match

---
 modules/datastore_volumes_info.pm | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/modules/datastore_volumes_info.pm b/modules/datastore_volumes_info.pm
index 55e8e71..53f7fe1 100644
--- a/modules/datastore_volumes_info.pm
+++ b/modules/datastore_volumes_info.pm
@@ -290,7 +290,7 @@ sub datastore_volumes_info
        $output = $tmp_output;
        }
 
-    if ($output)
+    if ($tmp_output or $tmp_output_error)
        {
        if ( $state == 0 )
           {
@@ -324,7 +324,12 @@ sub datastore_volumes_info
        }
     else
        {
-       if ($alertonly)
+       if(!$tmp_output and !$tmp_output_error)
+           {
+           $output = "No matching volumes found";
+           $state = 1;
+           }
+       elsif($alertonly)
           {
           $output = "OK. There are no alerts";
           }
