From 0ff8cab3ffa205322cfec78cd6cd7b5400c74486 Mon Sep 17 00:00:00 2001
From: Matthias Gallinger <matthias.gallinger@consol.de>
Date: Tue, 21 May 2019 13:36:06 +0200
Subject: [PATCH] skip storage checks in maintenance mode with ok

---
 modules/host_storage_info.pm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/modules/host_storage_info.pm b/modules/host_storage_info.pm
index 13abe47..805253e 100644
--- a/modules/host_storage_info.pm
+++ b/modules/host_storage_info.pm
@@ -84,7 +84,7 @@ sub host_storage_info
     if (($host_view->get_property('runtime.inMaintenanceMode')) eq "true")
        {
        print "Notice: " . $host_view->name . " is in maintenance mode, check skipped\n";
-       exit 1;
+       exit 0;
        }
    
     if (!defined($subselect))

