diff -Naur a/share/nagvis/htdocs/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo/Compiler.php b/share/nagvis/htdocs/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo/Compiler.php
--- a/share/nagvis/htdocs/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo/Compiler.php 2017-12-21 19:17:31.000000000 +0100
+++ b/share/nagvis/htdocs/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo/Compiler.php 2017-12-21 19:18:32.000000000 +0100
@@ -2434,7 +2434,7 @@
                }

                $breaker = false;
-               while (list($k,$char) = each($breakChars)) {
+               foreach ($breakChars AS $k => $char) {
                        $test = strpos($substr, $char);
                        if ($test !== false && $test < $end) {
                                $end = $test;
@@ -2939,7 +2939,7 @@
                }

                // loops over the param map and assigns values from the template or default value for unset optional params
-               while (list($k,$v) = each($map)) {
+               foreach ($map AS $k => $v) {
                        if ($v[0] === '*') {
                                // "rest" array parameter, fill every remaining params in it and then break
                                if (count($ps) === 0) {
