--- a/share/frontend/nagvis-js/ext/dwoo-1.1.0/plugins/builtin/blocks/textformat.php	2017-10-25 16:41:31.000000000 +0200
+++ b/share/frontend/nagvis-js/ext/dwoo-1.1.0/plugins/builtin/blocks/textformat.php	2017-12-22 07:23:25.000000000 +0100
@@ -67,7 +67,7 @@
 		// gets paragraphs
 		$pgs = explode("\n", str_replace(array("\r\n", "\r"), "\n", $this->buffer));
 
-		while (list($i,) = each($pgs)) {
+		foreach ($pgs AS $i => $j) {
 			if (empty($pgs[$i])) {
 				continue;
 			}
