--- a/share/frontend/nagvis-js/ext/dwoo-1.1.0/plugins/builtin/functions/capitalize.php	2017-10-25 16:41:31.000000000 +0200
+++ b/share/frontend/nagvis-js/ext/dwoo-1.1.0/plugins/builtin/functions/capitalize.php	2017-12-22 07:26:46.000000000 +0100
@@ -25,7 +25,7 @@
 	} else {
 		$bits = explode(' ', (string) $value);
 		$out = '';
-		while (list(,$v) = each($bits)) {
+		foreach ($bits AS $k => $v) {
 			if (preg_match('#^[^0-9]+$#', $v)) {
 				$out .=	' '.mb_convert_case($v, MB_CASE_TITLE, $dwoo->getCharset());
 			} else {
