--- a/share/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo.php	2017-10-25 16:41:31.000000000 +0200
+++ b/share/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo.php	2017-12-22 07:30:36.000000000 +0100
@@ -1193,7 +1193,7 @@
 		}
 		unset($varstr);
 
-		while (list($k, $sep) = each($m[1])) {
+		foreach ($m[1] AS $k => $sep) {
 			if ($sep === '.' || $sep === '[' || $sep === '') {
 				if ((is_array($data) || $data instanceof ArrayAccess) && ($safeRead === false || isset($data[$m[2][$k]]))) {
 					$data = $data[$m[2][$k]];
@@ -1365,7 +1365,7 @@
 			$cur = $this->scope;
 		}
 
-		while (list($k, $sep) = each($m[1])) {
+		foreach ($m[1] AS $k => $sep) {
 			if ($sep === '.' || $sep === '[' || $sep === '') {
 				if ((is_array($cur) || $cur instanceof ArrayAccess) && isset($cur[$m[2][$k]])) {
 					$cur = $cur[$m[2][$k]];
@@ -1410,7 +1410,7 @@
 			$cur =& $this->scope;
 			$last = array(array_pop($m[1]), array_pop($m[2]));
 
-			while (list($k, $sep) = each($m[1])) {
+			foreach ($m[1] AS $k => $sep) {
 				if ($sep === '.' || $sep === '[' || $sep === '') {
 					if (is_array($cur) === false) {
 						$cur = array();
