--- a/share/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo/Data.php	2017-10-25 16:41:31.000000000 +0200
+++ b/share/frontend/nagvis-js/ext/dwoo-1.1.0/Dwoo/Data.php	2017-12-22 07:33:33.000000000 +0100
@@ -71,7 +71,7 @@
 	public function mergeData(array $data)
 	{
 		$args = func_get_args();
-		while (list(,$v) = each($args)) {
+		foreach ($args AS $k => $v) {
 			if (is_array($v)) {
 				$this->data = array_merge($this->data, $v);
 			}
@@ -90,7 +90,7 @@
 	{
 		if (is_array($name)) {
 			reset($name);
-			while (list($k,$v) = each($name))
+			foreach ($name AS $k => $v) {
 				$this->data[$k] = $v;
 		} else {
 			$this->data[$name] = $val;
