--- pnp4nagios-0.6.42.orig/src/npcdmod.c	2025-11-10 07:43:40.457772077 +0000
+++ pnp4nagios-0.6.42/src/npcdmod.c	2025-11-10 07:53:57.351591941 +0000
@@ -52,7 +52,7 @@
 
 /* for backwards compatibilty */
 static void write_to_all_logs(char *buffer, int level) {
-	nm_log(level, buffer);
+	nm_log(level, "%s", buffer);
 }
 
 /* this function gets called when the module is loaded by the event broker */
