diff --git a/lib/Thruk.pm b/lib/Thruk.pm
index 52e3424..d40df04 100644
--- a/lib/Thruk.pm
+++ b/lib/Thruk.pm
@@ -167,8 +167,8 @@ BEGIN {
         # static content included?
         if(-d $addon.'root') {
             my $target_symlink = $project_root.'/root/thruk/plugins/'.$addon_name;
-            if(-e $target_symlink) { unlink($target_symlink) or die("cannot unlink: ".$target_symlink." : $!"); }
-            symlink($addon.'root', $target_symlink) or die("cannot create ".$target_symlink." : ".$!);
+            #if(-e $target_symlink) { unlink($target_symlink) or die("cannot unlink: ".$target_symlink." : $!"); }
+            #symlink($addon.'root', $target_symlink) or die("cannot create ".$target_symlink." : ".$!);
         }
     }
 }
