diff --git a/script/thruk_fastcgi.pl b/script/thruk_fastcgi.pl
index 013f5d9..a8e0003 100755
--- a/script/thruk_fastcgi.pl
+++ b/script/thruk_fastcgi.pl
@@ -6,7 +6,12 @@ use warnings;
 BEGIN {
     use FindBin;
     use Config;
-    if(-e $FindBin::Bin."/../local-lib") {
+    # omd path is different
+    if( -d "../../../lib/perl5/lib") {
+        use lib "../../../lib/perl5/lib/perl5";
+        require local::lib; local::lib->import("../../../lib/perl5/");
+    }
+    elsif(-e $FindBin::Bin."/../local-lib") {
         ### use 5.10.0; # only required for packages ###
         use lib $FindBin::Bin."/../local-lib/lib/perl5";
         if(! -e $FindBin::Bin."/../local-lib/lib/perl5/".$Config{archname}) {
