diff -Naur a/src/ctlib/ct.c b/src/ctlib/ct.c
--- a/src/ctlib/ct.c	2013-08-22 19:05:52.000000000 +0200
+++ b/src/ctlib/ct.c	2013-08-22 19:50:46.000000000 +0200
@@ -615,8 +615,13 @@
 		con->tds_socket = NULL;
 		return CS_FAIL;
 	}
-	if (con->server_addr)
+	if (con->server_addr) {
+		char ip_addr[256];
 		tds_dstr_copy(&connection->server_host_name, con->server_addr);
+		if (tds_lookup_host(tds_dstr_cstr(&connection->server_host_name), ip_addr) == TDS_SUCCEED) {
+			tds_dstr_copy(&connection->ip_addr, ip_addr);
+                }
+	}
 
 	/* override locale settings with CS_CONNECTION settings, if any */
 	if (con->locale) {
