--- a/share/pnp/application/vendor/fpdf/fpdf.php	2015-01-03 14:09:06.000000000 +0100
+++ b/share/pnp/application/vendor/fpdf/fpdf.php	2016-07-01 11:11:37.777163176 +0200
@@ -73,6 +73,10 @@
 *                               Public methods                                 *
 *                                                                              *
 *******************************************************************************/
+function __construct($orientation, $unit, $size) {
+	return(FPDF::FPDF($orientation, $unit, $size));
+}
+
 function FPDF($orientation='P', $unit='mm', $size='A4')
 {
 	// Some checks
@@ -879,7 +883,7 @@
 		$this->y += $h;
 }
 
-function Image($file, $x=null, $y=null, $w=0, $h=0, $type='', $link='')
+function Image($file, $x=null, $y=null, $w=0, $h=0, $type='', $link='', $align, $resize, $dpi, $palign, $ismask, $imgmask, $border)
 {
 	// Put an image on the page
 	if(!isset($this->images[$file]))
--- a/share/pnp/application/vendor/fpdf/fpdi_pdf_parser.php	2015-01-03 14:09:06.000000000 +0100
+++ b/share/pnp/application/vendor/fpdf/fpdi_pdf_parser.php	2016-07-01 11:10:20.070329174 +0200
@@ -66,6 +66,9 @@
      * @param string $filename  Source-Filename
      * @param object $fpdi      Object of type fpdi
      */
+	function __construct($filename, &$fpdi) {
+		return(fpdi_pdf_parser::fpdi_pdf_parser($filename, $fpdi));
+	}
     function fpdi_pdf_parser($filename, &$fpdi) {
         $this->fpdi =& $fpdi;
 		
--- a/share/pnp/application/vendor/fpdf/pdf_context.php	2015-01-03 14:09:06.000000000 +0100
+++ b/share/pnp/application/vendor/fpdf/pdf_context.php	2016-07-01 11:10:47.235905173 +0200
@@ -35,6 +35,9 @@
 
 	// Constructor
 
+	function __construct(&$f) {
+		return(pdf_context::pdf_context($f));
+	}
 	function pdf_context(&$f) {
 		$this->file =& $f;
 		if (is_string($this->file))
@@ -94,4 +97,4 @@
 	        return false;
 		}
 	}
-}
\ No newline at end of file
+}
--- a/share/pnp/application/vendor/fpdf/pdf_parser.php	2015-01-03 14:09:06.000000000 +0100
+++ b/share/pnp/application/vendor/fpdf/pdf_parser.php	2016-07-01 11:11:14.485523172 +0200
@@ -91,6 +91,9 @@
      *
      * @param string $filename  Source-Filename
      */
+	function __construct(&$f) {
+		return(pdf_parser::pdf_parser($f));
+	}
 	function pdf_parser($filename) {
         $this->filename = $filename;
         
