diff -Naur a/share/pnp/application/lib/json.php b/share/pnp/application/lib/json.php
--- a/share/pnp/application/lib/json.php	2015-01-03 14:09:06.000000000 +0100
+++ b/share/pnp/application/lib/json.php	2017-11-14 20:36:42.349373000 +0100
@@ -130,7 +130,7 @@
     *                                   bubble up with an error, so all return values
     *                                   from encode() should be checked with isError()
     */
-    function Services_JSON($use = 0)
+    function __construct($use = 0)
     {
         $this->use = $use;
     }
@@ -780,7 +780,7 @@
 
     class Services_JSON_Error extends PEAR_Error
     {
-        function Services_JSON_Error($message = 'unknown error', $code = null,
+        function __construct($message = 'unknown error', $code = null,
                                      $mode = null, $options = null, $userinfo = null)
         {
             parent::PEAR_Error($message, $code, $mode, $options, $userinfo);
@@ -794,7 +794,7 @@
      */
     class Services_JSON_Error
     {
-        function Services_JSON_Error($message = 'unknown error', $code = null,
+        function __construct($message = 'unknown error', $code = null,
                                      $mode = null, $options = null, $userinfo = null)
         {
 
