diff -Naur a/share/pnp/application/lib/json.php b/share/pnp/application/lib/json.php
--- a/share/pnp/application/models/data.php	2015-01-03 14:09:06.000000000 +0100
+++ b/share/pnp/application/models/data.php	2017-11-14 20:36:42.349373000 +0100
@@ -976,7 +976,7 @@
     */
     public function getFirstPage(){
         $pages = $this->getPages();
-        if(sizeof($pages) > 0 ){
+        if(is_array($pages) && sizeof($pages) > 0 ){
             return urldecode($pages[0]);
         }else{
             return FALSE; 
