diff -Naur a/scripts/process_perfdata.pl.in b/scripts/process_perfdata.pl.in
--- a/scripts/process_perfdata.pl.in	2015-01-03 14:09:06.000000000 +0100
+++ b/scripts/process_perfdata.pl.in	2018-02-16 10:08:33.324524300 +0100
@@ -304,6 +304,16 @@
         }
     }
 
+    # remove unprintable and control characters
+    $NAGIOS{PERFDATA} =~ s/[^[:ascii:]]//g;
+    $NAGIOS{PERFDATA} =~ s/[\x00-\x1f]//g;
+    if ($NAGIOS{SERVICEPERFDATA}) {
+      $NAGIOS{SERVICEPERFDATA} =~ s/[^[:ascii:]]//g;
+      $NAGIOS{SERVICEPERFDATA} =~ s/[\x00-\x1f]//g;
+    } elsif ($NAGIOS{HOSTPERFDATA}) {
+      $NAGIOS{HOSTPERFDATA} =~ s/[^[:ascii:]]//g;
+      $NAGIOS{HOSTPERFDATA} =~ s/[\x00-\x1f]//g;
+    }
     if ( $NAGIOS{PERFDATA} =~ /^(.*)\s\[(.*)\]$/ ) {
         $NAGIOS{PERFDATA}      = $1;
         $NAGIOS{CHECK_COMMAND} = $2;
