--- a/share/pnp/index.php.in	2022-08-02 16:29:59.677492923 +0200
+++ b/share/pnp/index.php.in	2022-08-02 16:30:50.587324686 +0200
@@ -31,7 +31,7 @@
 if(substr($_SERVER["SCRIPT_FILENAME"], 0, 4) == '/omd') {
     define('OMD', TRUE);
     $site_parts = array_slice(explode('/' ,dirname($_SERVER["SCRIPT_FILENAME"])), 0, -3);
-    define('OMD_SITE_ROOT', join($site_parts, '/'));
+    define('OMD_SITE_ROOT', join('/', $site_parts));
     define('OMD_SITE', $site_parts[count($site_parts)-1]);
     unset($site_parts);
 }else{
