--- a/lib/kohana/system/core/utf8.php	2022-08-02 20:48:08.755477711 +0200
+++ b/lib/kohana/system/core/utf8.php	2022-08-02 20:48:21.323945443 +0200
@@ -49,7 +49,7 @@
 	);
 }
 
-if (extension_loaded('mbstring') AND (ini_get('mbstring.func_overload') & MB_OVERLOAD_STRING))
+if (extension_loaded('mbstring') AND defined('MB_OVERLOAD_STRING') AND (ini_get('mbstring.func_overload') & MB_OVERLOAD_STRING))
 {
 	trigger_error
 	(
