--- a/lib/kohana/system/core/Kohana.php	2022-08-22 10:19:59.824073973 +0200
+++ b/lib/kohana/system/core/Kohana.php	2022-08-22 10:20:35.033374073 +0200
@@ -857,8 +857,8 @@
 	{
 		try
 		{
-			// PHP errors have 5 args, always
-			$PHP_ERROR = (func_num_args() === 5);
+			// PHP errors have 4 or 5 args, always
+			$PHP_ERROR = (func_num_args() === 4 || func_num_args() === 5);
 	
 			// Test to see if errors should be displayed
 			if ($PHP_ERROR AND (error_reporting() & $exception) === 0)
