--- a/share/pnp/application/models/data.php	2022-08-22 13:05:49.091693140 +0200
+++ b/share/pnp/application/models/data.php	2022-08-22 13:08:48.482659474 +0200
@@ -473,9 +473,9 @@
             $template_file = $this->findTemplate( $template, $type );
             $TIMERANGE     = $this->TIMERANGE;
         }
-        $def     = FALSE;
-        $opt     = FALSE;
-        $ds_name = FALSE;
+        $def     = [];
+        $opt     = [];
+        $ds_name = [];
         /*
         * 0.4.x Template compatibility 
         */
@@ -513,17 +513,17 @@
             $ds_name = $tmp;
         }
         //
-        if($def != FALSE){
+        if(count($def) > 0){
             $this->RRD['def'] = $this->array_reindex($def);
         }else{
             throw new Kohana_Exception('error.template-without-def', $template_file);
         }
-        if($opt != FALSE ){
+        if(count($opt) > 0){
             $this->RRD['opt'] = $this->array_reindex($opt);
         }else{
             throw new Kohana_Exception('error.template-without-opt', $template_file);
         }
-        if( $ds_name != FALSE ){
+        if(count($ds_name) > 0){
             $this->RRD['ds_name'] = $this->array_reindex($ds_name);
         }
         return TRUE;        
