--- a/share/pnp/application/controllers/mobile.php	2022-08-22 19:49:38.269592274 +0200
+++ b/share/pnp/application/controllers/mobile.php	2022-08-22 19:50:16.811002608 +0200
@@ -46,7 +46,7 @@
         $this->template->query = $this->add_view('mobile_search');
         $query     = pnp::clean($this->input->post('term'));
         $result    = array();
-        if(strlen($query)>=1) {
+        if($query !== NULL && strlen($query)>=1) {
             $hosts = $this->data->getHosts();
             foreach($hosts as $host){
                 if(preg_match("/$query/i",$host['name'])){
