--- a/lib/kohana/system/libraries/Controller.php	2023-03-30 15:38:50.173865593 +0200
+++ b/lib/kohana/system/libraries/Controller.php	2023-03-30 15:39:02.354322283 +0200
@@ -11,6 +11,8 @@
  * @license    http://kohanaphp.com/license.html
  */
 abstract class Controller_Core {
+    public $uri   = NULL;
+    public $input = NULL;
 
 	// Allow all controllers to run in production by default
 	const ALLOW_PRODUCTION = TRUE;
