--- a/share/pnp/application/models/auth.php	2023-03-30 15:43:55.877326444 +0200
+++ b/share/pnp/application/models/auth.php	2023-03-30 15:44:10.913890106 +0200
@@ -16,6 +16,8 @@
     public $AUTH_ENABLED = FALSE;
     public $REMOTE_USER = NULL;
 
+    public $config       = NULL;
+
     public function __construct() {
         $this->config = new Config_Model;
         $this->config->read_config();
