--- a/share/pnp/application/controllers/zoom.php	2023-03-30 23:02:35.457023900 +0200
+++ b/share/pnp/application/controllers/zoom.php	2023-03-30 23:04:13.596643426 +0200
@@ -7,7 +7,9 @@
  * @license    GPL
  */
 class Zoom_Controller extends System_Controller  {
-
+    public $graph_width = NULL;
+    public $graph_height = NULL;
+    public $url = NULL;
 
     public function __construct()
     {
