--- a/share/pnp/templates.dist/check_multi.php	2022-08-08 08:59:56.477596327 +0200
+++ b/share/pnp/templates.dist/check_multi.php	2022-08-08 09:00:37.807091043 +0200
@@ -4,7 +4,7 @@
 # Template for check_multi
 # 
 
-$opt[1] = "--lower=$MIN[1] --vertical-label num  --title \"Number of Checks\" ";
+$opt[1] = "--lower-limit=$MIN[1] --vertical-label num  --title \"Number of Checks\" ";
 $ds_name[1] = "Executed Plugins";
 
 $def[1] =  "DEF:var1=$RRDFILE[1]:$DS[1]:AVERAGE " ;
