--- a/lib/plugins/formular/syntax.php	2016-07-25 18:23:09.466220446 +0200
+++ b/lib/plugins/formular/syntax.php	2016-07-25 18:23:24.714677839 +0200
@@ -115,12 +115,12 @@
 		require_once DOKU_INC . 'inc/parser/xhtml.php';
 		require_once DOKU_INC . 'inc/parser/parser.php';
 		
-		$Parser = & new Doku_Parser(); 
-		$Parser->Handler = & new Doku_Handler();
+		$Parser = new Doku_Parser(); 
+		$Parser->Handler = new Doku_Handler();
 		$Parser->addMode('php',new Doku_Parser_Mode_PHP());
 		
 		$instructions = $Parser->parse($string);		// Get a list of instructions
-		$Renderer = & new Doku_Renderer_XHTML();		// Create a renderer
+		$Renderer = new Doku_Renderer_XHTML();		// Create a renderer
 		
 		// Loop through the instructions
 		foreach ( $instructions as $instruction ) {
@@ -134,8 +134,8 @@
 		require_once DOKU_INC . 'inc/parser/xhtml.php';
 		require_once DOKU_INC . 'inc/parser/parser.php';
 		
-		$Parser = & new Doku_Parser(); 
-		$Parser->Handler = & new Doku_Handler();
+		$Parser = new Doku_Parser(); 
+		$Parser->Handler = new Doku_Handler();
 		
 		$Parser->addMode('listblock',new Doku_Parser_Mode_ListBlock());
 		$Parser->addMode('preformatted',new Doku_Parser_Mode_Preformatted()); 
@@ -171,7 +171,7 @@
 		}
 		
 		$instructions = $Parser->parse($string);		// Get a list of instructions
-		$Renderer = & new Doku_Renderer_XHTML();		// Create a renderer
+		$Renderer = new Doku_Renderer_XHTML();		// Create a renderer
 		
 		// Loop through the instructions
 		foreach ( $instructions as $instruction ) {
