--- a/lib/plugins/formular/syntax.php	2022-08-10 10:07:32.224437467 +0200
+++ b/lib/plugins/formular/syntax.php	2022-08-10 10:07:29.501437467 +0200
@@ -677,7 +677,7 @@
 	/**
 	* Handle the match
 	*/
-	function handle($match, $state, $pos, &$handler) {
+	function handle($match, $state, $pos, $handler) {
 		global $ID;
 		
 		switch ($state) {
@@ -1084,7 +1084,7 @@
 	/**
 	 * Create output
 	 */
-	function render($mode, &$renderer, $data) {
+	function render($mode, $renderer, $data) {
 		if ($mode == 'xhtml') {
             $renderer->info['cache'] = false;
 			$renderer->doc .= $data;
