--- a/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Image/ImageProcessor.php	2022-08-11 16:05:45.701119246 +0200
+++ b/lib/plugins/dw2pdf/vendor/mpdf/mpdf/src/Image/ImageProcessor.php	2022-08-11 16:06:09.458013902 +0200
@@ -531,7 +531,7 @@
 										if ($colorindex >= $n) {
 											$alpha = 255;
 										} else {
-											$alpha = ord($transparency{$colorindex});
+											$alpha = ord($transparency[$colorindex]);
 										} // 0-255
 										if ($alpha > 0) {
 											imagesetpixel($imgalpha, $xpx, $ypx, $alpha);
@@ -1039,7 +1039,7 @@
 								if ($colorindex >= $n) {
 									$alpha = 255;
 								} else {
-									$alpha = ord($transparency{$colorindex});
+									$alpha = ord($transparency[$colorindex]);
 								} // 0-255
 								$mimgdata .= chr($alpha);
 							}
