From 24095bf9be070142f50f3e7be3995c7ec5765325 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Thu, 13 Mar 2025 19:42:25 +0100
Subject: [PATCH 1/2] trim whitespace at end of plugin output

---
 src/naemon/checks.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/naemon/checks.c b/src/naemon/checks.c
index 6f4a423e9..13a07e65e 100644
--- a/src/naemon/checks.c
+++ b/src/naemon/checks.c
@@ -315,6 +315,9 @@ int process_check_result(check_result *cr)
 
 	source_name = check_result_source(cr);
 
+	/* trim whitespace at the end of the plugin output */
+	rstrip(cr->output);
+
 	if (cr->object_check_type == SERVICE_CHECK) {
 		service *svc;
 		svc = find_service(cr->host_name, cr->service_description);

From 7020408749a088700813ac44b2b0b48fe8ae181d Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Thu, 13 Mar 2025 20:20:23 +0100
Subject: [PATCH 2/2] remove backslashes before semicolons

since a couple of versions it is possible to escape semicolons with backslashes in config files.

For example:

    check_command         check_dummy!0!'test |val=5c\;\;\;0\;'

But naemon did not remove the backslashes so they remain in the command line. This commit
fixes this and removes backslashes before semicolons.
---
 src/naemon/xodtemplate.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/naemon/xodtemplate.c b/src/naemon/xodtemplate.c
index f7461d04a..4c5a642a4 100644
--- a/src/naemon/xodtemplate.c
+++ b/src/naemon/xodtemplate.c
@@ -8319,6 +8319,7 @@ static int xodtemplate_process_config_file(char *filename)
 	register int x = 0;
 	register int y = 0;
 	char *ptr = NULL;
+	int has_escaped_semicolon = 0;
 
 
 	if (verify_config >= 2)
@@ -8342,6 +8343,7 @@ static int xodtemplate_process_config_file(char *filename)
 	while (1) {
 
 		nm_free(inputbuf);
+		has_escaped_semicolon = 0;
 
 		/* read the next line */
 		if(use_precached_objects) {
@@ -8358,6 +8360,7 @@ static int xodtemplate_process_config_file(char *filename)
 						break;
 					else if (inputbuf[x - 1] != '\\')
 						break;
+					has_escaped_semicolon = 1;
 				}
 			}
 			inputbuf[x] = '\x0';
@@ -8372,6 +8375,17 @@ static int xodtemplate_process_config_file(char *filename)
 		if (input[0] == '\x0' || input[0] == '#')
 			continue;
 
+		/* remove backslashes before semicolons */
+		if(has_escaped_semicolon) {
+			for (x = 0; input[x + 1] != '\x0'; x++) {
+				if (input[x] == '\\' && input[x + 1] == ';') {
+					for (y = x; input[y] != '\x0'; y++)
+						input[y] = input[y + 1];
+					input[y] = '\x0';
+				}
+			}
+		}
+
 		/* this is the start of an object definition */
 		if (!strncmp(input, "define", 6)) {
 
