From 45d9bf7ce2741cce0794d11c20e88cb713cdedb0 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Tue, 25 Mar 2025 12:37:01 +0100
Subject: [PATCH] fix crash in check post processing

Signed-off-by: Sven Nierlein <sven@consol.de>
---
 src/naemon/checks.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/naemon/checks.c b/src/naemon/checks.c
index 72679bf0..b33fa4a3 100644
--- a/src/naemon/checks.c
+++ b/src/naemon/checks.c
@@ -317,7 +317,8 @@ int process_check_result(check_result *cr)
 	source_name = check_result_source(cr);
 
 	/* trim whitespace at the end of the plugin output */
-	rstrip(cr->output);
+	if(cr->output != NULL)
+		rstrip(cr->output);
 
 	if (cr->object_check_type == SERVICE_CHECK) {
 		service *svc;
