diff -Naur naemon-1.4.1.orig/src/naemon/notifications.h naemon-1.4.1/src/naemon/notifications.h
--- naemon-1.4.1.orig/src/naemon/notifications.h	2018-05-30 12:46:04.000000000 +0000
+++ naemon-1.4.1/src/naemon/notifications.h	2023-11-14 11:46:39.458885429 +0000
@@ -83,7 +83,7 @@
 	struct notify_list *next;
 } notification;
 
-const char *notification_reason_name(unsigned int reason_type);
+const char *notification_reason_name(enum NotificationReason reason);
 int check_service_notification_viability(service *, int, int);			/* checks viability of notifying all contacts about a service */
 int is_valid_escalation_for_service_notification(service *, serviceescalation *, int);	/* checks if an escalation entry is valid for a particular service notification */
 int should_service_notification_be_escalated(service *);			/* checks if a service notification should be escalated */
