--- a/src/github.com/grafana/grafana/pkg/api/index.go    2016-04-04 11:28:28.728000000 +0200
+++ b/src/github.com/grafana/grafana/pkg/api/index.go    2016-04-04 11:28:06.448000000 +0200
@@ -47,7 +47,7 @@
        if themeUrlParam == "light" {
                data.User.LightTheme = true
        }
-       data.User.LightTheme = true
+
        dashboardChildNavs := []*dtos.NavLink{
                {Text: "Home", Url: setting.AppSubUrl + "/"},
                {Text: "Playlists", Url: setting.AppSubUrl + "/playlists"},
