--- a/configure	2011-02-07 11:04:20.000000000 +0100
+++ b/configure	2011-02-23 23:12:36.000000000 +0100
@@ -5288,17 +5288,17 @@
 user="$with_user"
 
 
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking if user $user exists" >&5
-$as_echo_n "checking if user $user exists... " >&6; }
-id -a $user > /dev/null 2>&1
-if test "$?" != "0" ; then
-    { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
-$as_echo "no" >&6; }
-    as_fn_error "user $user does not exist" "$LINENO" 5
-else
-    { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-$as_echo "yes" >&6; }
-fi
+#{ $as_echo "$as_me:${as_lineno-$LINENO}: checking if user $user exists" >&5
+#$as_echo_n "checking if user $user exists... " >&6; }
+#id -a $user > /dev/null 2>&1
+#if test "$?" != "0" ; then
+#    { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
+#$as_echo "no" >&6; }
+#    as_fn_error "user $user does not exist" "$LINENO" 5
+#else
+#    { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
+#$as_echo "yes" >&6; }
+#fi
 
 ##############################################
 # Check some functions
