From 6efd40ba0737e7c0a7feaf3d766fec836f30ef63 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@nierlein.de>
Date: Tue, 26 Sep 2017 11:09:37 +0200
Subject: [PATCH] fix memory leak when using naemon

---
 Changes                    | 1 +
 neb_module/result_thread.c | 6 ++++--
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/neb_module/result_thread.c b/neb_module/result_thread.c
index f7181d2..ec38f13 100644
--- a/neb_module/result_thread.c
+++ b/neb_module/result_thread.c
@@ -33,7 +33,9 @@ static const char *gearman_worker_source_name(void *source) {
     if(!source)
         return "unknown internal source (voodoo, perhaps?)";
 
-    return (char*) source;
+    // we cannot return the source here as it would be never freed
+    //return (char*) source;
+    return "Mod-Gearman Worker";
 }
 
 struct check_engine mod_gearman_check_engine = {
@@ -226,7 +228,7 @@ void *get_results( gearman_job_st *job, void *context, size_t *result_size, gear
             chk_result->service_description = gm_strdup( value );
         } else if ( !strcmp( key, "source" ) ) {
 #ifdef USENAEMON
-            chk_result->source = gm_strdup( value );
+            chk_result->source = value;
 #endif
         } else if ( !strcmp( key, "check_options" ) ) {
             chk_result->check_options = atoi( value );
-- 
2.1.4

