--- a/Makefile.am	2011-02-22 15:29:23.000000000 +0100
+++ b/Makefile.am	2011-02-22 15:29:48.000000000 +0100
@@ -152,12 +152,7 @@
 
 install-local-state-dir:
 	$(RM) $(DESTDIR)$(pkglibdir)/mod_gearman.so
-	if [ `id -u` -eq 0 ]; then \
-	    $(install_sh_PROGRAM) -o $(user) -d $(DESTDIR)$(localstatedir)/mod_gearman/; \
-    else \
-	    $(install_sh_PROGRAM) -o $(user) -d $(DESTDIR)$(localstatedir)/mod_gearman/; \
-        echo "*** You should run: sudo chown $(user) $(DESTDIR)$(localstatedir)/mod_gearman/"; \
-    fi
+	$(install_sh_PROGRAM) -d $(DESTDIR)$(localstatedir)/mod_gearman/
 
 install-data-local: install-local-state-dir
 	$(install_sh_PROGRAM) -m 644 extras/shared.conf            $(DESTDIR)$(datadir)/mod_gearman/shared.conf
