From 3c7833b1c0289e8ee313a55855940bba5f60d543 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Wed, 1 Mar 2023 12:58:14 +0100
Subject: [PATCH] fix deadlock if logfile is not writeable on startup

---
 neb_module_naemon/mod_gearman.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/neb_module_naemon/mod_gearman.c b/neb_module_naemon/mod_gearman.c
index 5d6e1f9..ddd222f 100644
--- a/neb_module_naemon/mod_gearman.c
+++ b/neb_module_naemon/mod_gearman.c
@@ -1163,7 +1163,9 @@ void init_logging(mod_gm_opt_t *opt) {
     if(opt->logmode == GM_LOG_MODE_FILE && opt->logfile && opt->debug_level < GM_LOG_STDOUT) {
         opt->logfile_fp = fopen(opt->logfile, "a+");
         if(opt->logfile_fp == NULL) {
+            opt->logmode = GM_LOG_MODE_CORE;
             gm_log( GM_LOG_ERROR, "error opening logfile: %s\n", opt->logfile );
+            opt->logmode = GM_LOG_MODE_FILE;
         }
     }
     if ( opt->logmode == GM_LOG_MODE_AUTO ) {
