From 2910d663548952527020ce53e93ea85c29dfbfd6 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Fri, 23 Feb 2024 21:48:49 +0100
Subject: [PATCH] fix sending performance data into multiple queues

---
 neb_module_naemon/mod_gearman.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/neb_module_naemon/mod_gearman.c b/neb_module_naemon/mod_gearman.c
index 60499b5..dbbc408 100644
--- a/neb_module_naemon/mod_gearman.c
+++ b/neb_module_naemon/mod_gearman.c
@@ -1451,8 +1451,6 @@ int handle_perfdata(int event_type, void *data) {
                     break;
                 }
 
-                make_uniq(uniq, "%s", hostchkdata->host_name);
-
                 /* replace newlines with actual newlines */
                 perf_data = replace_str(hostchkdata->perf_data, "\\n", "\n");
 
@@ -1491,8 +1489,6 @@ int handle_perfdata(int event_type, void *data) {
                     break;
                 }
 
-                make_uniq(uniq, "%s-%s", srvchkdata->host_name, srvchkdata->service_description);
-
                 /* replace newlines with actual newlines */
                 perf_data = replace_str(srvchkdata->perf_data, "\\n", "\n");
 
@@ -1526,6 +1522,19 @@ int handle_perfdata(int event_type, void *data) {
         int i = 0;
         for (i = 0; i < mod_gm_opt->perfdata_queues_num; i++) {
             char *perfdata_queue = mod_gm_opt->perfdata_queues_list[i];
+
+            /* generate uuid including the queue name. it seems like pushing the same uuid into different queues still overwrites them. */
+            if(mod_gm_opt->perfdata_mode == GM_PERFDATA_OVERWRITE) {
+                switch (event_type) {
+                    case NEBCALLBACK_HOST_CHECK_DATA:
+                        make_uniq(uniq, "%s-%s", perfdata_queue, hostchkdata->host_name);
+                        break;
+                    case NEBCALLBACK_SERVICE_CHECK_DATA:
+                        make_uniq(uniq, "%s-%s-%s", perfdata_queue, srvchkdata->host_name, srvchkdata->service_description);
+                        break;
+                }
+            }
+
             /* add our job onto the queue */
             if(add_job_to_queue(&client,
                                  mod_gm_opt->server_list,
