From 5241bdc985ec0fa7e63bc9a37f7d8238ec7b7c9d Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Tue, 3 Mar 2026 16:07:28 +0100
Subject: [PATCH] fix duplicating notes_url when expanding with unclosed macros

---
 src/OffsetStringMacroColumn.cc | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/OffsetStringMacroColumn.cc b/src/OffsetStringMacroColumn.cc
index de5749b..8cccd43 100644
--- a/src/OffsetStringMacroColumn.cc
+++ b/src/OffsetStringMacroColumn.cc
@@ -7,7 +7,6 @@
 #include "OffsetStringMacroColumn.h"
 #include "AndingFilter.h"
 #include "Query.h"
-#include "logger.h"
 #include "nagios.h"
 extern char     *macro_user[MAX_USER_MACROS];
 
@@ -30,7 +29,7 @@ string OffsetStringMacroColumn::valueAsString(void *data, Query *)
         result += string(scan, dollar - scan);
         const char *otherdollar = strchr(dollar + 1, '$');
         if (!otherdollar) { // unterminated macro, do not expand
-            result += scan;
+            result += dollar; // add the remaining text, including the dollar sign
             break;
         }
         string macroname = string(dollar + 1, otherdollar - dollar - 1);
