From e7c526cd2ae8442a96c1708134f709fc1d33e418 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Tue, 3 Mar 2026 16:10:49 +0100
Subject: [PATCH] add check_timeout attributes for hosts/services

---
 src/TableHosts.cc    | 2 ++
 src/TableServices.cc | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/src/TableHosts.cc b/src/TableHosts.cc
index 3c9e67a..803a5dc 100644
--- a/src/TableHosts.cc
+++ b/src/TableHosts.cc
@@ -74,6 +74,8 @@ void TableHosts::addColumns(Table *table, string prefix, int indirect_offset)
                 "IP address", (char *)(&hst.address) - ref, indirect_offset));
     table->addColumn(new OffsetStringColumn(prefix + "check_command",
                 "Naemon command for active host check of this host", (char *)(&hst.check_command) - ref, indirect_offset));
+    table->addColumn(new OffsetIntColumn(prefix + "check_timeout",
+                "Host check timeout", (char *)(&hst.check_timeout) - ref, indirect_offset));
     table->addColumn(new OffsetStringColumn(prefix + "event_handler",
                 "Naemon command used as event handler", (char *)(&hst.event_handler) - ref, indirect_offset));
     table->addColumn(new OffsetStringColumn(prefix + "notification_period",
diff --git a/src/TableServices.cc b/src/TableServices.cc
index 245bfe1..721b727 100644
--- a/src/TableServices.cc
+++ b/src/TableServices.cc
@@ -208,6 +208,8 @@ void TableServices::addColumns(Table *table, string prefix, int indirect_offset,
                 "An optional display name", (char *)(&svc.display_name) - ref, indirect_offset));
     table->addColumn(new OffsetStringColumn(prefix + "check_command",
                 "Naemon command used for active checks", (char *)(&svc.check_command) - ref, indirect_offset));
+    table->addColumn(new OffsetIntColumn(prefix + "check_timeout",
+                "Service check timeout", (char *)(&svc.check_timeout) - ref, indirect_offset));
     table->addColumn(new OffsetStringColumn(prefix + "event_handler",
                 "Naemon command used as event handler", (char *)(&svc.event_handler) - ref, indirect_offset));
     table->addColumn(new OffsetStringColumn(prefix + "plugin_output",
