From 0a46e473ea81a6e96bf5f238b45818d00fb3e6fd Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Mon, 4 Nov 2024 20:09:15 +0100
Subject: [PATCH 1/2] remove unused function

---
 src/module.c | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/src/module.c b/src/module.c
index b29d07e..a0bdd7f 100644
--- a/src/module.c
+++ b/src/module.c
@@ -771,20 +771,6 @@ void livestatus_parse_arguments(const char *args_orig)
     free(tmp);
 }
 
-void omd_advertize()
-{
-    char *omd_site = getenv("OMD_SITE");
-    if (omd_site) {
-        if (g_debug_level > 0)
-            logger(LG_INFO, "Running on OMD site %s. Cool.", omd_site);
-    }
-    else {
-        logger(LG_INFO, "Hint: please try out OMD - the Open Monitoring Distribution");
-        logger(LG_INFO, "Please visit OMD at http://omdistro.org");
-    }
-}
-
-
 /* this function gets called when the module is loaded by the event broker */
 int nebmodule_init(int flags __attribute__ ((__unused__)), char *args, void *handle)
 {

From b466f0d53daadca496533df8c3dcd2c367397c4b Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@consol.de>
Date: Mon, 4 Nov 2024 21:19:25 +0100
Subject: [PATCH 2/2] fix crash when requesting invalid table name

for example when doing http request accidentially on the livestatus socket:

    curl -vv --max-time 1 --unix-socket /omd/sites/dev/tmp/run/live http://test
---
 src/Store.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Store.cc b/src/Store.cc
index 73790ec..cbd4c2e 100644
--- a/src/Store.cc
+++ b/src/Store.cc
@@ -189,7 +189,7 @@ void Store::answerGetRequest(InputBuffer *input, OutputBuffer *output, const cha
     }
     Query query(input, output, table);
 
-    if(table->hasLogcache()) {
+    if(table && table->hasLogcache()) {
         g_store->logCache()->lockLogCache();
         // check if client is still connected, we might have waited too long for the lock
         if(!output->isAlive(fd)) {
