--- a/mysql.xs
+++ b/mysql.xs
@@ -790,7 +790,7 @@ dbd_mysql_get_info(dbh, sql_info_type)
     D_imp_dbh(dbh);
     IV type = 0;
     SV* retsv=NULL;
-#if !defined(MARIADB_BASE_VERSION) && MYSQL_VERSION_ID >= 50709
+#if MYSQL_VERSION_ID >= 50709
 /* MariaDB 10 is not MySQL source level compatible so this only applies to MySQL*/
     IV buffer_len;
 #endif 
@@ -822,7 +822,7 @@ dbd_mysql_get_info(dbh, sql_info_type)
 	    retsv = newSVpvn("`", 1);
 	    break;
 	case SQL_MAXIMUM_STATEMENT_LENGTH:
-#if !defined(MARIADB_BASE_VERSION) && MYSQL_VERSION_ID >= 50709
+#if MYSQL_VERSION_ID >= 50709
         /* MariaDB 10 is not MySQL source level compatible so this
            only applies to MySQL*/
 	    /* mysql_get_option() was added in mysql 5.7.3 */
