--- a/dbdimp.h
+++ b/dbdimp.h
@@ -20,6 +20,9 @@
 #include <DBIXS.h>  /* installed by the DBI module                        */
 #include <mysql.h>  /* Comes with MySQL-devel */
 #include <mysqld_error.h>  /* Comes MySQL */
+#if defined(MARIADB_BASE_VERSION) && MYSQL_VERSION_ID >= 100210
+#include <mariadb_version.h> /* Comes with MariaDB Connector C */
+#endif
 
 #include <errmsg.h> /* Comes with MySQL-devel */
 
