From ceb94c5296ac9a7a7cd10bf87aa6f98aa8f3bc77 Mon Sep 17 00:00:00 2001
From: Paul Miller <jettero@corky.vhb.voltar.org>
Date: Thu, 11 Jan 2018 09:46:36 -0500
Subject: [PATCH] this is really the only change I actually want apparently

---
 dbdimp.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dbdimp.c b/dbdimp.c
index 9b8b313..06fce86 100644
--- a/dbdimp.c
+++ b/dbdimp.c
@@ -1997,7 +1997,13 @@ MYSQL *mysql_dr_connect(
         we turn off Mysql's auto reconnect and handle re-connecting ourselves
         so that we can keep track of when this happens.
       */
+
+#if MYSQL_VERSION_ID >= 50013
+      my_bool reconnect=false;
+      mysql_options(result, MYSQL_OPT_RECONNECT, &reconnect);
+#else
       result->reconnect=0;
+#endif
     }
     else {
       /* 
