--- a/plugins/utils.h
+++ b/plugins/utils.h
@@ -200,10 +200,10 @@
 #define DBG_PRINT(x, ...)                                                                          \
 	do {                                                                                           \
 		if (verbose > x) {                                                                         \
-			printf(0 __VA_OPT__(, ) __VA_ARGS__);                                                  \
+			printf(__VA_ARGS__);                                                                  \
 		}                                                                                          \
 	} while (0);
 
-#define DBG_PRINT_1(...) DBG_PRINT(1, 0 __VA_OPT__(, ) __VA_ARGS__);
-#define DBG_PRINT_2(...) DBG_PRINT(2, 0 __VA_OPT__(, ) __VA_ARGS__);
-#define DBG_PRINT_3(...) DBG_PRINT(3, 0 __VA_OPT__(, ) __VA_ARGS__);
+#define DBG_PRINT_1(...) DBG_PRINT(1, __VA_ARGS__);
+#define DBG_PRINT_2(...) DBG_PRINT(2, __VA_ARGS__);
+#define DBG_PRINT_3(...) DBG_PRINT(3, __VA_ARGS__);
