--- a/plugins/check_multi.in	2017-08-16 14:45:17.747301025 +0200
+++ b/plugins/check_multi.in	2017-08-16 14:47:00.115360700 +0200
@@ -1347,7 +1347,7 @@
 	}
 	while (<OBJECTS_CACHE>) {
 		#--- begin of object, determine type
-		if (/^define ([a-z0-9\_\-]+) {$/) {
+		if (/^define ([a-z0-9\_\-]+) \{$/) {
 			$type="$1";
 			$typelist{$type}++;
 			$objectcount++;
@@ -4136,7 +4136,7 @@
 
 	#--- some debugging first
 	DEBUG4("MULTI Environment (sorted):\n\t".join("\n\t",get_env_vars('^MULTI')));
-	DEBUG4("${NAGIOS} Environment (sorted):\n\t".join("\n\t",get_env_vars('^${NAGIOS}')));
+	DEBUG4("${NAGIOS} Environment (sorted):\n\t".join("\n\t",get_env_vars('^$\{NAGIOS\}')));
 
 	#--- construction site for persistence
 	if ($opt{set}{test} && $opt{set}{persistent}) {
