--- nagvis-1.5.3/share/frontend/nagvis-js/classes/FrontendTemplateSystem.php.orig	2010-10-04 18:13:33.000000000 +0200
+++ nagvis-1.5.3/share/frontend/nagvis-js/classes/FrontendTemplateSystem.php	2010-10-04 18:14:25.000000000 +0200
@@ -33,9 +33,11 @@
 		$this->CORE = $CORE;
 		
 		// Load Dwoo. It is used as external library
-		require_once($this->CORE->getMainCfg()->getValue('paths','base').'share/frontend/nagvis-js/ext/dwoo-1.1.0/dwooAutoload.php');
+		require_once($this->CORE->getMainCfg()->getValue('paths','base')
+		             .HTDOCS_DIR.'/frontend/nagvis-js/ext/dwoo-1.1.0/dwooAutoload.php');
 
-		$this->TMPL = new Dwoo($this->CORE->getMainCfg()->getValue('paths','var').'tmpl/compile', $this->CORE->getMainCfg()->getValue('paths','var').'tmpl/cache');
+		$this->TMPL = new Dwoo($this->CORE->getMainCfg()->getValue('paths','var')
+		                       .'tmpl/compile', $this->CORE->getMainCfg()->getValue('paths','var').'tmpl/cache');
 	}
 	
 	public function getTmplSys() {
