--- nagvis-1.5.2/share/server/core/defines/global.php.orig	2010-09-29 23:42:27.000000000 +0200
+++ nagvis-1.5.2/share/server/core/defines/global.php	2010-09-29 23:43:12.000000000 +0200
@@ -58,7 +58,7 @@
 define('DEBUGLEVEL', 6);
 
 // Path to the debug file
-define('DEBUGFILE', '../../../var/nagvis-debug.log');
+define('DEBUGFILE', join(array_slice(explode('/' ,dirname($_SERVER["SCRIPT_FILENAME"])), 0, -5), '/').'/tmp/nagvis/nagvis-debug.log');
 
 // It is possible to define a second main configuration file
 // to pre-define some options in a file the user may not be
@@ -72,11 +72,11 @@
 // The last value wins.
 //
 // Path to the main configuration file
-define('CONST_MAINCFG', '../../../etc/nagvis.ini.php');
-define('CONST_MAINCFG_CACHE', '../../../var/nagvis.ini.php');
+define('CONST_MAINCFG', join(array_slice(explode('/' ,dirname($_SERVER["SCRIPT_FILENAME"])), 0, -5), '/').'/etc/nagvis/nagvis.ini.php');
+define('CONST_MAINCFG_CACHE', join(array_slice(explode('/' ,dirname($_SERVER["SCRIPT_FILENAME"])), 0, -5), '/').'/tmp/nagvis/nagvis.ini.php');
 
 // Path to "site" main configuration file
-//define('CONST_MAINCFG_SITE', '../../../etc/nagvis-site.ini.php');
+define('CONST_MAINCFG_SITE', join(array_slice(explode('/' ,dirname($_SERVER["SCRIPT_FILENAME"])), 0, -5), '/').'/etc/nagvis/nagvis-omd.ini.php');
 
 // Needed minimal PHP version
 define('CONST_NEEDED_PHP_VERSION', '5.0');
