--- main.go	2017-11-20 13:34:16.000000000 +0100
+++ main.go	2018-02-21 10:01:01.000000000 +0100
@@ -19,6 +19,7 @@ import (
 	"github.com/prometheus/common/version"
 
 	"github.com/prometheus/snmp_exporter/config"
+    "strconv"
 )
 
 var (
@@ -72,6 +73,18 @@ func handler(w http.ResponseWriter, r *h
 	}
 	log.Debugf("Scraping target '%s' with module '%s'", target, moduleName)
 
+    snmpVersion := r.URL.Query().Get("version")
+    if v, err := strconv.Atoi(snmpVersion); snmpVersion != "" && err != nil {
+        log.Debugf("Overriding version. Old version: '%d' new version: '%d'", module.Version, v)
+        module.Version = v
+    }
+
+    community := r.URL.Query().Get("community")
+    if community != "" {
+        log.Debugf("Overriding community. Old community: '%s' new community: '%s'", module.Auth.Community, community)
+        module.Auth.Community = community
+    }
+
 	start := time.Now()
 	registry := prometheus.NewRegistry()
 	collector := collector{target: target, module: module}
