--- go/src/github.com/prometheus/snmp_exporter/main.go	2020-02-24 09:41:32.000000000 +0100
+++ go/src/github.com/prometheus/snmp_exporter/main.go	2020-02-24 15:42:28.699410065 +0100
@@ -33,6 +33,8 @@
 	"gopkg.in/alecthomas/kingpin.v2"
 	yaml "gopkg.in/yaml.v2"
 
+	"strconv"
+
 	"github.com/prometheus/snmp_exporter/config"
 )
 
@@ -86,6 +88,7 @@
 	if moduleName == "" {
 		moduleName = "if_mib"
 	}
+
 	sc.RLock()
 	module, ok := (*(sc.C))[moduleName]
 	sc.RUnlock()
@@ -95,6 +98,18 @@
 		return
 	}
 
+	snmpVersion := query.Get("version")
+	if v, err := strconv.Atoi(snmpVersion); snmpVersion != "" && err == nil {
+		level.Debug(logger).Log("msg", "Overriding version", "versions", v)
+		module.WalkParams.Version = v
+	}
+
+	community := query.Get("community")
+	if community != "" {
+		level.Debug(logger).Log("msg", "Overriding community", "community", community)
+		module.WalkParams.Auth.Community = config.Secret(community)
+	}
+
 	logger = log.With(logger, "module", moduleName, "target", target)
 	level.Debug(logger).Log("msg", "Starting scrape")
 
