--- go/src/github.com/prometheus/snmp_exporter/main.go	2018-12-07 14:52:10.000000000 +0100
+++ go/src/github.com/prometheus/snmp_exporter/main.go	2018-12-07 14:58:28.942116918 +0100
@@ -32,6 +32,7 @@
 	"github.com/prometheus/common/version"
 
 	"github.com/prometheus/snmp_exporter/config"
+   "strconv"
 )
 
 var (
@@ -65,6 +66,7 @@
 	prometheus.MustRegister(version.NewCollector("snmp_exporter"))
 }
 
+
 func handler(w http.ResponseWriter, r *http.Request) {
 	target := r.URL.Query().Get("target")
 	if target == "" {
@@ -84,6 +86,18 @@
 		snmpRequestErrors.Inc()
 		return
 	}
+    snmpVersion := r.URL.Query().Get("version")
+    if v, err := strconv.Atoi(snmpVersion); snmpVersion != "" && err != nil {
+            log.Debugf("Overriding version. Old version: '%d' new version: '%d'", module.WalkParams.Version, v)
+            module.WalkParams.Version = v
+    }   
+
+    community := r.URL.Query().Get("community")
+    if community != "" {
+        log.Debugf("Overriding community. Old community: '%s' new community: '%s'", module.WalkParams.Auth.Community, community)
+        module.WalkParams.Auth.Community = config.Secret(community)
+    }
+
 	log.Debugf("Scraping target '%s' with module '%s'", target, moduleName)
 
 	start := time.Now()
