--- a/histou/bootstrap.php	2022-08-10 14:50:00.694331726 +0200
+++ b/histou/bootstrap.php	2022-08-10 14:50:37.595742356 +0200
@@ -14,10 +14,12 @@
     return substr(realpath(dirname(__FILE__)), 0, -6);
 }
 
-function __autoload($className)
-{
+spl_autoload_register('histou_autoload');
+
+function histou_autoload($className) {
     $file = strtolower(str_replace('\\', DIRECTORY_SEPARATOR, $className)).'.php';
     if (file_exists($file)) {
         require_once $file;
     }
 }
+
