--- a/histou/database/influxdb.php	2020-06-29 10:59:54.000000000 +0200
+++ b/histou/database/influxdb.php	2023-01-31 17:42:33.194329665 +0100
@@ -38,13 +38,14 @@
     **/
     public function fetchPerfData()
     {
+        $seriesLimit = 100;
         $result = $this->makeGetRequest(
             sprintf(
-                "select * from metrics where host='%s' and service='%s' GROUP BY performanceLabel ORDER BY time DESC LIMIT 1",
+                "select * from metrics where host='%s' and service='%s' GROUP BY performanceLabel ORDER BY time DESC LIMIT 1 SLIMIT ".($seriesLimit + 1),
                 HOST,
                 SERVICE
             ).';'.sprintf(
-                "select * from metrics where host='%s' and service='%s' GROUP BY performanceLabel LIMIT 1",
+                "select * from metrics where host='%s' and service='%s' GROUP BY performanceLabel LIMIT 1 SLIMIT ".($seriesLimit + 1),
                 HOST,
                 SERVICE
             )
@@ -54,6 +55,10 @@
         } elseif (empty($result['results'][0])) {
             return $result['results'][1];
         } else {
+            if(count($result['results'][0]["series"]) > $seriesLimit) {
+                $result['results'][0]["series"][$seriesLimit]["tags"]["performanceLabel"] = "{ERROR: MORE THAN $seriesLimit SERIES FOUND}"; # add brackets so it appears as last item
+                $result['results'][0]["series"][$seriesLimit]["values"][0][1] = ""; # command name
+            }
             return $result['results'][0];
         }
     }
