From 2fedf18b6c1f2e9500e6f9213d0c213e7df5b23e Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@nierlein.de>
Date: Tue, 5 Jun 2018 11:03:16 +0200
Subject: [PATCH] sakuli: use legend url parameter if available

---
 templates/default/check_sakuli.php | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/templates/default/check_sakuli.php b/templates/default/check_sakuli.php
index ec1489a..b5852cd 100644
--- a/templates/default/check_sakuli.php	2018-06-05 11:09:38.910179065 +0200
+++ b/templates/default/check_sakuli.php	2018-06-05 11:09:44.678192972 +0200
@@ -13,7 +13,7 @@
     $host = '.*',
     $service = '.*',
     $command = '.*',
-    $perfLabel = array('^c_\d\d\d_|^s_\d\d\d_\d\d\d_|^suite_')
+    $perfLabel = array('^c_\d\d\d_|^s_\d\d\d_|^suite_|^graph_type_')
 );
 
 $genTemplate = function ($perfData) {
@@ -25,7 +25,7 @@
     $dashboard->addDefaultAnnotations($perfData['host'], $perfData['service']);
     $dashboard->addAnnotation(
         "errors",
-        "SELECT path,service FROM images WHERE host = '".$perfData['host']."' AND service = '".$perfData['service']."'",
+        "SELECT path,service FROM images WHERE host = '".$perfData['host']."' AND service = '".$perfData['service']."' AND \$timeFilter ORDER BY time DESC LIMIT 100",
         "path",
         "service",
         "",
@@ -48,6 +48,9 @@
     $caseNames = array();
     $steps = array();
     $maxSteps = 0;
+    $graphType = "";
+    $stepName = "";
+    $stepID = "";
     foreach ($perfData['perfLabel'] as $key => $values) {
         if (preg_match(';suite_([^_].*);', $key, $hit)) {
             $suiteLabel = $hit[0];
@@ -59,7 +62,7 @@
             array_push($caseNames, $hit[1]);
             continue;
         }
-        if (preg_match(';s_(\d\d\d)_[^_].*;', $key, $hit)) {
+        if (preg_match(';s_(\d\d\d)_([^_].*);', $key, $hit)) {
             if (!array_key_exists($hit[1], $steps)) {
                 $steps[$hit[1]] = 0;
             }
@@ -67,99 +70,129 @@
             if ($steps[$hit[1]] > $maxSteps) {
                 $maxSteps = $steps[$hit[1]];
             }
+            # only used for step dashboard. There is only one step.
+            $stepId = $hit[1];
+            $stepName = $hit[2];
+            continue;
+        }
+        if (preg_match(';c_\d\d\d_([^_].*);', $key, $hit)) {
+            array_push($caseLabels, $hit[0]);
+            array_push($caseNames, $hit[1]);
+            continue;
+        }
+        if (preg_match(';graph_type_(.*);', $key, $hit)) {
+            $graphType = $hit[1];
             continue;
         }
     }
 
 
     //Suite Row
-    $suiteRow = new \histou\grafana\Row("Suite-Runtime");
-    //Suite Panel
-    $suitePanel = \histou\grafana\graphpanel\GraphPanelFactory::generatePanel($perfData['service']." (Sakuli suite $suiteName) on ".$perfData['host']);
-    $suitePanel->setLeftUnit("s");
-    $suitePanel->setSpan(11);
-    $suitePanel->setLeftYAxisMinMax(0);
-    $suitePanel->fillBelowLine("/^(?!warning|critical).*$/", 5);
-    $suitePanel->setLegend(true, true, true, true, false, false, true, true, true, true);
-    $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], $suiteLabel, $caseColors[0], $suiteName));
-    $c = 1;
-    for ($i = 0; $i < sizeof($caseLabels); $i++) {
-        $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], $caseLabels[$i], $caseColors[$c], $caseNames[$i]));
-        $suitePanel->stack($caseNames[$i].'-value');
-        $c = ($c + 1) % (sizeof($caseColors) - 1);
-    }
-    $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 'suite__warning', '#FFFC15', 'warning'));
-    $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 'suite__critical', '#FF3727', 'critical'));
-    $suiteRow->addPanel($suitePanel);
-    //Suite Stat Panel
-    $suiteStat = \histou\grafana\singlestatpanel\SinglestatPanelFactory::generatePanel("");
-    $suiteStat->setSpan(1);
-    $suiteStat->addTarget($suiteStat->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 'suite__state'));
-    $suiteStat->setColor(array('#99ff66', '#ffc125', '#ee0000'));
-    $suiteStat->setThresholds("1", "2");
-    $suiteStat->addRangeToTextElement(0, 0.5, 'OK');
-    $suiteStat->addRangeToTextElement(0.5, 1.5, 'Warn');
-    $suiteStat->addRangeToTextElement(1.5, 2.5, 'Crit');
-    $suiteStat->addRangeToTextElement(2.5, 3.5, 'Unkn');
-    $suiteRow->addPanel($suiteStat);
+    if (preg_match(';^(suite|)$;', $graphType)) {
+        $suiteRow = new \histou\grafana\Row("Suite runtime");
+        //Suite Panel
+        $suitePanel = \histou\grafana\graphpanel\GraphPanelFactory::generatePanel("Sakuli suite $suiteName (Host: ".$perfData['host'].")");
+        $suitePanel->setLeftUnit("s");
+        $suitePanel->setSpan(11);
+        $suitePanel->setLeftYAxisMinMax(0);
+        $suitePanel->fillBelowLine("/^(?!warning|critical).*$/", 5);
+        $suitePanel->setLegend(SHOW_LEGEND, true, true, true, false, false, true, true, true, true);
+        $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], $suiteLabel, $caseColors[0], $suiteName));
+        $c = 1;
+        for ($i = 0; $i < sizeof($caseLabels); $i++) {
+            $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], $caseLabels[$i], $caseColors[$c], $caseNames[$i]));
+            $suitePanel->stack($caseNames[$i].'-value');
+            $c = ($c + 1) % (sizeof($caseColors) - 1);
+        }
+        $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 'suite__warning', '#FFFC15', 'warning'));
+        $suitePanel->addTarget($suitePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 'suite__critical', '#FF3727', 'critical'));
+        $suiteRow->addPanel($suitePanel);
+        //Suite Stat Panel
+        $suiteStat = \histou\grafana\singlestatpanel\SinglestatPanelFactory::generatePanel("");
+        $suiteStat->setSpan(1);
+        $suiteStat->addTarget($suiteStat->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 'suite__state'));
+        $suiteStat->setColor(array('#99ff66', '#ffc125', '#ee0000'));
+        $suiteStat->setThresholds("1", "2");
+        $suiteStat->addRangeToTextElement(0, 0.5, 'OK');
+        $suiteStat->addRangeToTextElement(0.5, 1.5, 'Warn');
+        $suiteStat->addRangeToTextElement(1.5, 2.5, 'Crit');
+        $suiteStat->addRangeToTextElement(2.5, 3.5, 'Unkn');
+        $suiteRow->addPanel($suiteStat);
 
-    $dashboard->addRow($suiteRow);
+        $dashboard->addRow($suiteRow);
+    }
 
     //Case Row
-    $caseRow = new \histou\grafana\Row("Case-Runtime");
-    $caseRow->setCustomProperty("repeat", $templateName);
-    //Case Graph Panel
-    $casePanel = \histou\grafana\graphpanel\GraphPanelFactory::generatePanel($perfData['service']." (Sakuli case #$templateVariableString) on ".$perfData['host']);
-    $casePanel->setLeftUnit("s");
-    $casePanel->setSpan(11);
-    $casePanel->setLeftYAxisMinMax(0);
-    $casePanel->fillBelowLine("/^(?!warning|critical).*$/", 5);
-    $casePanel->setLegend(true, true, true, true, false, false, true, true, true, true);
-    $target1 = $casePanel->createTarget(array('host' => array('value' => $perfData['host'], 'operator' => '='),
-                                    'service' => array('value' => $perfData['service'], 'operator' => '='),
-                                    'command' => array('value' => $perfData['command'], 'operator' => '='),
-                                    'performanceLabel' => array('value' => \histou\helper\str::genRegex('c_'.$templateVariableString.'_[^_].*'), 'operator' => '=~')
-                                    ));
-    $target1 = $casePanel->addXToTarget($target1, array('value'), '', '');
-    $target1['alias'] = '$tag_performanceLabel';
-    $target1['groupBy'] = array( array("params"=>array("\$interval"), "type"=> "time"),
-                                 array("params"=>array("performanceLabel"), "type"=> "tag"),
-                                 array("params"=>array("null"), "type"=> "fill"));
-    $casePanel->addTarget($target1);
-    $target = $casePanel->createTarget(array('host' => array('value' => $perfData['host'], 'operator' => '='),
-                                    'service' => array('value' => $perfData['service'], 'operator' => '='),
-                                    'command' => array('value' => $perfData['command'], 'operator' => '='),
-                                    'performanceLabel' => array('value' => \histou\helper\str::genRegex('s_'.$templateVariableString.'.*'), 'operator' => '=~')
-                                    ));
-    $target = $casePanel->addXToTarget($target, array('value'), '', '');
-    $target['alias'] = '$tag_performanceLabel';
-    $target['groupBy'] = array(     array("params"=>array("\$interval"), "type"=> "time"),
-                                array("params"=>array("performanceLabel"), "type"=> "tag"),
-                                array("params"=>array("null"), "type"=> "fill"));
-    $casePanel->addTarget($target);
-    $casePanel->stack('/s_\d\d\d_.*/');
-    $casePanel->addTarget($casePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], "c_$templateVariableString"."__warning", '#FFFC15', 'warning'));
-    $casePanel->addTarget($casePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], "c_$templateVariableString"."__critical", '#FF3727', 'critical'));
-    $c = 0;
-    for ($i = 0; $i < $maxSteps; $i++) {
-        $casePanel->addRegexColor(sprintf('/s_\d\d\d_%03d_.*/', $i), $stepColors[$c]);
-        $c = ($c + 1) % (sizeof($stepColors) - 1);
-    }
-    $caseRow->addPanel($casePanel);
+    if (preg_match(';^(suite|case)$;', $graphType)) {
+        $caseRow = new \histou\grafana\Row("Case runtime");
+        $caseRow->setCustomProperty("repeat", $templateName);
+        //Case Graph Panel
+        $casePanel = \histou\grafana\graphpanel\GraphPanelFactory::generatePanel("Sakuli case #$templateVariableString (suite $suiteName) on ".$perfData['host']);
+        $casePanel->setLeftUnit("s");
+        $casePanel->setSpan(11);
+        $casePanel->setLeftYAxisMinMax(0);
+        $casePanel->fillBelowLine("/^(?!warning|critical).*$/", 5);
+        $casePanel->setLegend(SHOW_LEGEND, true, true, true, false, false, true, true, true, true);
+        $target1 = $casePanel->createTarget(array('host' => array('value' => $perfData['host'], 'operator' => '='),
+                                        'service' => array('value' => $perfData['service'], 'operator' => '='),
+                                        'command' => array('value' => $perfData['command'], 'operator' => '='),
+                                        'performanceLabel' => array('value' => \histou\helper\str::genRegex('c_'.$templateVariableString.'_[^_].*'), 'operator' => '=~')
+                                        ));
+        $target1 = $casePanel->addXToTarget($target1, array('value'), '', '');
+        $target1['alias'] = '$tag_performanceLabel';
+        $target1['groupBy'] = array( array("params"=>array("\$interval"), "type"=> "time"),
+                                     array("params"=>array("performanceLabel"), "type"=> "tag"),
+                                     array("params"=>array("null"), "type"=> "fill"));
+        $casePanel->addTarget($target1);
+        $target = $casePanel->createTarget(array('host' => array('value' => $perfData['host'], 'operator' => '='),
+                                        'service' => array('value' => $perfData['service'], 'operator' => '='),
+                                        'command' => array('value' => $perfData['command'], 'operator' => '='),
+                                        'performanceLabel' => array('value' => \histou\helper\str::genRegex('s_'.$templateVariableString.'.*'), 'operator' => '=~')
+                                        ));
+        $target = $casePanel->addXToTarget($target, array('value'), '', '');
+        $target['alias'] = '$tag_performanceLabel';
+        $target['groupBy'] = array(     array("params"=>array("\$interval"), "type"=> "time"),
+                                    array("params"=>array("performanceLabel"), "type"=> "tag"),
+                                    array("params"=>array("null"), "type"=> "fill"));
+        $casePanel->addTarget($target);
+        $casePanel->stack('/s_\d\d\d_.*/');
+        $casePanel->addTarget($casePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], "c_$templateVariableString"."__warning", '#FFFC15', 'warning'));
+        $casePanel->addTarget($casePanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], "c_$templateVariableString"."__critical", '#FF3727', 'critical'));
+        $c = 0;
+        for ($i = 0; $i < $maxSteps; $i++) {
+            $casePanel->addRegexColor(sprintf('/s_\d\d\d_%03d_.*/', $i), $stepColors[$c]);
+            $c = ($c + 1) % (sizeof($stepColors) - 1);
+        }
+        $caseRow->addPanel($casePanel);
 
-    //Case Stat Panel
-    $caseStat = \histou\grafana\singlestatpanel\SinglestatPanelFactory::generatePanel("");
-    $caseStat->setSpan(1);
-    $caseStat->addTarget($caseStat->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], "c_$templateVariableString".'__state'));
-    $caseStat->setColor(array('#99ff66', '#ffc125', '#ee0000'));
-    $caseStat->setThresholds("1", "2");
-    $caseStat->addRangeToTextElement(0, 0.5, 'OK');
-    $caseStat->addRangeToTextElement(0.5, 1.5, 'Warn');
-    $caseStat->addRangeToTextElement(1.5, 2.5, 'Crit');
-    $caseStat->addRangeToTextElement(2.5, 3.5, 'Unkn');
-    $caseRow->addPanel($caseStat);
+        //Case Stat Panel
+        $caseStat = \histou\grafana\singlestatpanel\SinglestatPanelFactory::generatePanel("");
+        $caseStat->setSpan(1);
+        $caseStat->addTarget($caseStat->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], "c_$templateVariableString".'__state'));
+        $caseStat->setColor(array('#99ff66', '#ffc125', '#ee0000'));
+        $caseStat->setThresholds("1", "2");
+        $caseStat->addRangeToTextElement(0, 0.5, 'OK');
+        $caseStat->addRangeToTextElement(0.5, 1.5, 'Warn');
+        $caseStat->addRangeToTextElement(1.5, 2.5, 'Crit');
+        $caseStat->addRangeToTextElement(2.5, 3.5, 'Unkn');
+        $caseRow->addPanel($caseStat);
 
-    $dashboard->addRow($caseRow);
+        $dashboard->addRow($caseRow);
+    }
 
+    //Step Row
+    if (preg_match(';^step$;', $graphType)) {
+        $stepRow = new \histou\grafana\Row("Step runtime");
+        //Step Graph Panel
+        $stepPanel = \histou\grafana\graphpanel\GraphPanelFactory::generatePanel($perfData['service']." (Sakuli step $stepName, suite $suiteName) on ".$perfData['host']);
+        $stepPanel->setLeftUnit("s");
+        $stepPanel->setSpan(12);
+        $stepPanel->setLeftYAxisMinMax(0);
+        $stepPanel->fillBelowLine("/^(?!warning|critical).*$/", 5);
+        $stepPanel->setLegend(SHOW_LEGEND, true, true, true, false, false, true, true, true, true);
+        $stepPanel->addTarget($stepPanel->genTargetSimple($perfData['host'], $perfData['service'], $perfData['command'], 's_'.$stepId.'_'.$stepName, $stepColors[0], $stepName));
+        $stepRow->addPanel($stepPanel);
+        $dashboard->addRow($stepRow);
+    }
     return $dashboard;
 };
