--- a/plugins-root/check_icmp.c	2023-10-18 11:11:23.000000000 +0200
+++ b/plugins-root/check_icmp.c	2023-10-18 11:24:33.269915331 +0200
@@ -1313,7 +1313,9 @@
 			 * conspicuously as missing entries in perfparse and cacti */
 			pl = 100;
 			rta = 0;
-			status = STATE_CRITICAL;
+			if (pl >= crit.pl)
+                            status = STATE_CRITICAL;
+
 			/* up the down counter if not already counted */
 			if(!(host->flags & FLAG_LOST_CAUSE) && targets_alive) targets_down++;
 		} else {
@@ -1449,7 +1451,7 @@
 
 
 	/* this is inevitable */
-	if(!targets_alive) status = STATE_CRITICAL;
+	if(!targets_alive && crit.pl <= 100) status = STATE_CRITICAL;
 	if(min_hosts_alive > -1) {
 		if(hosts_ok >= min_hosts_alive) status = STATE_OK;
 		else if((hosts_ok + hosts_warn) >= min_hosts_alive) status = STATE_WARNING;
@@ -1466,7 +1468,8 @@
 		}
 		i++;
 		if(!host->icmp_recv) {
-			status = STATE_CRITICAL;
+			if (pl >= crit.pl)
+				status = STATE_CRITICAL;
 			host->rtmin=0;
 			host->jitter_min=0;
 			if(host->flags & FLAG_LOST_CAUSE) {
@@ -1907,7 +1910,7 @@
 	if(!th->rta) return -1;
 
 	if(th->rta > MAXTTL * 1000000) th->rta = MAXTTL * 1000000;
-	if(th->pl > 100) th->pl = 100;
+	if(th->pl > 101) th->pl = 101;
 
 	return 0;
 }
