--- gearmand-0.25/util/instance.cc.orig 2011-10-28 22:33:38.000000000 -0500
+++ gearmand-0.25/util/instance.cc  2012-01-12 20:02:51.303519844 -0600
@@ -48,6 +48,7 @@
 #include <sys/socket.h>
 #include <sys/types.h>
 #include <netinet/in.h>
+#include <unistd.h>
 
 
 namespace datadifferential {
--- gearmand-0.25/util/pidfile.cc.orig  2011-09-26 23:59:09.000000000 -0500
+++ gearmand-0.25/util/pidfile.cc   2012-01-12 20:42:49.747797880 -0600
@@ -48,6 +48,7 @@
 #include <sstream>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <unistd.h>
 
 extern "C" {
 
--- gearmand-0.25/util/logfile.cc.orig  2011-09-26 23:59:09.000000000 -0500
+++ gearmand-0.25/util/logfile.cc   2012-01-12 20:57:25.526030905 -0600
@@ -49,6 +49,7 @@
 #include <sstream>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <unistd.h>
 
 namespace datadifferential {
 namespace util {
