From 2de4d33c670963a7dbc8c079eac41499fbe69bcb Mon Sep 17 00:00:00 2001
From: Ed Sabol <edwardjsabol@gmail.com>
Date: Mon, 29 Apr 2019 19:57:02 -0400
Subject: [PATCH] Fix strncpy() warning in libtest/socket.cc when compiling
 with gcc 8

---
 libtest/socket.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/libtest/socket.cc b/libtest/socket.cc
index 76ac5bf1..19d5a877 100644
--- a/libtest/socket.cc
+++ b/libtest/socket.cc
@@ -55,7 +55,8 @@ void set_default_socket(const char *socket)
 {
   if (socket)
   {
-    strncpy(global_socket, socket, strlen(socket));
+    strncpy(global_socket, socket, sizeof(global_socket));
+    global_socket[sizeof(global_socket) -1]= '\0';
   }
 }
 
